sporadic add owner migration failure

If you flush your docker volumes (as I continuously do), and run make, it will sporadically fail (once every 10 times, maybe) like this:

```
make migrate
make[1]: Entering directory '/home/alvaro/ft_transcendence'
docker-compose exec transcendence-app npm run knex:migrate:dev -w transcendence-app

> transcendence-app@0.0.1 knex:migrate:dev
> knex migrate:latest

Requiring external module ts-node/register
Using environment: development
migration file "20221206105801_add_web_owner_user.ts" failed
migration failed with error: ENOENT: no such file or directory, open '/var/lib/transcendence-app/avatars/b5bb33219cfd3f4ff858ba6724478e83'
ENOENT: no such file or directory, open '/var/lib/transcendence-app/avatars/b5bb33219cfd3f4ff858ba6724478e83'
Error: ENOENT: no such file or directory, open '/var/lib/transcendence-app/avatars/b5bb33219cfd3f4ff858ba6724478e83'
npm ERR! Lifecycle script `knex:migrate:dev` failed with error: 
npm ERR! Error: command failed 
npm ERR!   in workspace: transcendence-app@0.0.1 
npm ERR!   at location: /service/transcendence-app 
make[1]: *** [Makefile:92: migrate] Error 1
make[1]: Leaving directory '/home/alvaro/ft_transcendence'
make: *** [Makefile:10: all] Error 2

```
This is not often an issue, but may critically be an issue when we have corrections, since we have to have migrations, or a way around them, there as well. We should fix it.

I've ran make on a clean slate 6 times and it has failed twice. So my estimation was wron, this is more serious than expected. Last time, with these changes:
```
Index: transcendence-app/migrations/20221206105801_add_web_owner_user.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/transcendence-app/migrations/20221206105801_add_web_owner_user.ts b/transcendence-app/migrations/20221206105801_add_web_owner_user.ts
--- a/transcendence-app/migrations/20221206105801_add_web_owner_user.ts	(revision 67ca65d1362e47c3e74641dbe3a9a775f17ab38f)
+++ b/transcendence-app/migrations/20221206105801_add_web_owner_user.ts	(date 1674683544794)
@@ -18,15 +18,56 @@
 };
 
 export async function up(knex: Knex): Promise<void> {
-  const adminAvatar = await createRandomAvatar();
-  const adminWithPassword = await createAdminWithPassword(adminAvatar.id);
-  // Inserts admin entry
-  await knex('localfile').insert(adminAvatar);
-  const [admin, _] = await knex('users')
-    .insert(adminWithPassword)
-    .returning('*');
-  await knex('usertorole').insert({ id: admin.id, role: 'owner' });
-  await knex('usertorole').insert({ id: admin.id, role: 'moderator' });
+  try {
+    const adminAvatar = await createRandomAvatar();
+    try {
+      const adminWithPassword = await createAdminWithPassword(adminAvatar.id);
+      // Inserts admin entry
+      try {
+        await knex('localfile').insert(adminAvatar);
+        try {
+          const [admin, _] = await knex('users')
+            .insert(adminWithPassword)
+            .returning('*');
+          try {
+            await knex('usertorole').insert({ id: admin.id, role: 'owner' });
+            try {
+              await knex('usertorole').insert({
+                id: admin.id,
+                role: 'moderator',
+              });
+            } catch (e) {
+              console.log(
+                "failed at await knex('usertorole').insert({ id: admin.id, role: 'moderator' });\n",
+              );
+              throw e;
+            }
+          } catch (e) {
+            console.log(
+              "failed at await knex('usertorole').insert({ id: admin.id, role: 'owner' });",
+            );
+            throw e;
+          }
+        } catch (e) {
+          console.log(
+            "failed at await knex('users')\n" +
+              '    .insert(adminWithPassword)\n' +
+              "    .returning('*');",
+          );
+          throw e;
+        }
+      } catch (e) {
+        console.log("failed at knex('localfile').insert(adminAvatar)");
+        throw e;
+      }
+    } catch (e) {
+      console.log('failed at createAdminWithPassword');
+      throw e;
+    }
+  } catch (e) {
+    console.log('failed at create random avatar');
+    throw e;
+  }
 }
 
 export async function down(knex: Knex): Promise<void> {
```
I get these debugging lines:
```
make migrate
make[1]: Entering directory '/home/alvaro/ft_transcendence'
docker-compose exec transcendence-app npm run knex:migrate:dev -w transcendence-app

> transcendence-app@0.0.1 knex:migrate:dev
> knex migrate:latest

Requiring external module ts-node/register
Using environment: development
failed at create random avatar
migration file "20221206105801_add_web_owner_user.ts" failed
migration failed with error: ENOENT: no such file or directory, open '/var/lib/transcendence-app/avatars/a69ba753dd6bd00ac7d983980be3b19f'
ENOENT: no such file or directory, open '/var/lib/transcendence-app/avatars/a69ba753dd6bd00ac7d983980be3b19f'
Error: ENOENT: no such file or directory, open '/var/lib/transcendence-app/avatars/a69ba753dd6bd00ac7d983980be3b19f'
npm ERR! Lifecycle script `knex:migrate:dev` failed with error: 
npm ERR! Error: command failed 
npm ERR!   in workspace: transcendence-app@0.0.1 
npm ERR!   at location: /service/transcendence-app 
make[1]: *** [Makefile:92: migrate] Error 1
make[1]: Leaving directory '/home/alvaro/ft_transcendence'
make: *** [Makefile:10: all] Error 2
```
So the broken path is in the add avatar, and possibly localFileService. Which is confusing, because we're calling this method thousands of times in the seeding, and it was never complaining. Why should it break so frequently when adding a single user.

```
Requiring external module ts-node/register
Using environment: development
migration file "20221206105801_add_web_owner_user.ts" failed
migration failed with error: ENOENT: no such file or directory, open '/var/lib/transcendence-app/avatars/fdc6dcd399aa41789186533f0be9a3c5'
ENOENT: no such file or directory, open '/var/lib/transcendence-app/avatars/fdc6dcd399aa41789186533f0be9a3c5'
Error: ENOENT: no such file or directory, open '/var/lib/transcendence-app/avatars/fdc6dcd399aa41789186533f0be9a3c5'
npm ERR! Lifecycle script `knex:migrate:dev` failed with error: 
npm ERR! Error: command failed 
npm ERR!   in workspace: transcendence-app@0.0.1 
npm ERR!   at location: /service/transcendence-app 
make[1]: *** [Makefile:93: migrate] Error 1
make[1]: Leaving directory '/home/alvaro/ft_transcendence'
make: *** [Makefile:10: all] Error 2
```

Oh look, I've seen it again

Sadly, it reproduces remotely as well, [here](https://github.com/42AGV/ft_transcendence/actions/runs/4459619470/jobs/7832214491#step:4:3423) 
```
migrations_1  | Error: ENOENT: no such file or directory, open '/var/lib/transcendence-app/avatars/7e6a795e6096d0b29888f4f61a3fcd71'
migrations_1  | migration file "20221206105801_add_web_owner_user.ts" failed
migrations_1  | migration failed with error: ENOENT: no such file or directory, open '/var/lib/transcendence-app/avatars/7e6a795e6096d0b29888f4f61a3fcd71'
migrations_1  | ENOENT: no such file or directory, open '/var/lib/transcendence-app/avatars/7e6a795e6096d0b29888f4f61a3fcd71'
migrations_1  | Error: ENOENT: no such file or directory, open '/var/lib/transcendence-app/avatars/7e6a795e6096d0b29888f4f61a3fcd71'
e7a33230adc2   ft_transcendence/transcendence-app-dev   "/usr/bin/dumb-init …"   11 seconds ago   Up 10 seconds                        ft_transcendence_migrations_1
migrations_1  | npm ERR! Lifecycle script `knex:migrate:dev` failed with error: 
migrations_1  | npm ERR! Error: command failed 
migrations_1  | npm ERR!   in workspace: transcendence-app@0.0.1 
migrations_1  | npm ERR!   at location: /service/transcendence-app 
ft_transcendence_migrations_1 exited with code 1
```

--- a/transcendence-app/seeds/utils/index.ts
+++ b/transcendence-app/seeds/utils/index.ts
@@ -20,6 +20,9 @@ export const createRandomAvatar = async () => {
     localFileRepository,
   );
   const avatarDto = await localFileService.createRandomSVGFile(12, 512);
+  if (!avatarDto) {
+    throw new Error('Could not create a random avatar');
+  }
 
   return { id: uuidv4(), ...avatarDto };
 };

--- a/transcendence-app/src/auth/auth.controller.ts
+++ b/transcendence-app/src/auth/auth.controller.ts
@@ -106,6 +106,7 @@ export class AuthController {
   @ApiCreatedResponse({ description: 'Login successfully', type: User })
   @ApiBadRequestResponse({ description: 'Bad Request' })
   @ApiUnprocessableEntityResponse({ description: 'Unprocessable Entity' })
+  @ApiServiceUnavailableResponse({ description: 'Service Unavailable' })
   async registerLocalUser(
     @Body() registerUserDto: RegisterUserDto,
   ): Promise<User> {

--- a/transcendence-app/src/auth/auth.service.ts
+++ b/transcendence-app/src/auth/auth.service.ts
@@ -2,6 +2,7 @@ import {
   BadRequestException,
   ForbiddenException,
   Injectable,
+  ServiceUnavailableException,
   UnprocessableEntityException,
 } from '@nestjs/common';
 import { v4 as uuidv4 } from 'uuid';
@@ -60,6 +61,9 @@ export class AuthService {
     const { confirmationPassword: _, ...newUser } = user;
 
     const avatarDto = await this.localFileService.createRandomSVGFile(12, 512);
+    if (!avatarDto) {
+      throw new ServiceUnavailableException();
+    }
     const avatarId = uuidv4();
     const userDto: CreateUserDto = {
       ...newUser,

--- a/transcendence-app/src/chat/chat.controller.ts
+++ b/transcendence-app/src/chat/chat.controller.ts
@@ -80,6 +80,7 @@ export class ChatController {
   @ApiCreatedResponse({ description: 'Create a chatroom', type: Chatroom })
   @ApiBadRequestResponse({ description: 'Bad Request' })
   @ApiUnprocessableEntityResponse({ description: 'Unprocessable entity' })
+  @ApiServiceUnavailableResponse({ description: 'Service Unavailable' })
   async createChatroom(
     @GetUser() user: User,
     @Body() createChatDto: CreateChatroomDto,

--- a/transcendence-app/src/chat/chat.service.ts
+++ b/transcendence-app/src/chat/chat.service.ts
@@ -3,6 +3,7 @@ import {
   ForbiddenException,
   Injectable,
   NotFoundException,
+  ServiceUnavailableException,
 } from '@nestjs/common';
 import { v4 as uuidv4 } from 'uuid';
 import { BooleanString } from '../shared/enums/boolean-string.enum';
@@ -77,6 +78,9 @@ export class ChatService {
   async createChatroom(ownerId: string, chatroom: CreateChatroomDto) {
     const { confirmationPassword: _, ...newChat } = chatroom;
     const avatarDto = await this.localFileService.createRandomSVGFile(12, 512);
+    if (!avatarDto) {
+      throw new ServiceUnavailableException();
+    }
     const avatarId = uuidv4();
     if (chatroom.password) {
       if (chatroom.password !== chatroom.confirmationPassword) {

--- a/transcendence-app/src/shared/local-file/local-file.service.ts
+++ b/transcendence-app/src/shared/local-file/local-file.service.ts
@@ -3,12 +3,13 @@ import { LocalFileDto } from './local-file.dto';
 import { ConfigService } from '@nestjs/config';
 import { randomBytes } from 'crypto';
 import { join } from 'path';
-import { createWriteStream, mkdir, ReadStream, unlink } from 'fs';
+import { createWriteStream, ReadStream, unlink } from 'fs';
 import { LocalFileConfig } from './local-file.config.interface';
 import { LocalFile } from './infrastructure/db/local-file.entity';
 import { ILocalFileRepository } from './infrastructure/db/local-file.repository';
 import generateAvatar = require('github-like-avatar-generator');
 import { AVATARS_PATH } from '../constants';
+import { mkdir } from 'node:fs/promises';
 
 @Injectable()
 export class LocalFileService {
@@ -19,22 +20,24 @@ export class LocalFileService {
     private localFileRepository: ILocalFileRepository,
   ) {}
 
-  private createDirectory(directory: string) {
-    return mkdir(directory, { recursive: true }, (err) => {
-      if (err) {
+  private async createDirectory(directory: string) {
+    try {
+      await mkdir(directory, { recursive: true });
+    } catch (err: unknown) {
+      if (err instanceof Error) {
         this.logger.error(err.message);
-        throw err;
       }
-    });
+      throw err;
+    }
   }
 
-  private createWriteOutStream(path: string, mimetype: string) {
+  private async createWriteOutStream(path: string, mimetype: string) {
     const filesPath = this.configService.get(
       'TRANSCENDENCE_APP_DATA',
     ) as string;
     const filename = randomBytes(16).toString('hex');
     const directory = join(filesPath, path);
-    this.createDirectory(directory);
+    await this.createDirectory(directory);
     const finalPath = join(directory, filename);
     const outStream = createWriteStream(finalPath);
 
@@ -59,7 +62,7 @@ export class LocalFileService {
     path: string,
     mimetype: string,
   ): Promise<LocalFileDto> {
-    const { outStream, outStreamPromise } = this.createWriteOutStream(
+    const { outStream, outStreamPromise } = await this.createWriteOutStream(
       path,
       mimetype,
     );
@@ -74,7 +77,7 @@ export class LocalFileService {
     path: string,
     mimetype: string,
   ): Promise<LocalFileDto> {
-    const { outStream, outStreamPromise } = this.createWriteOutStream(
+    const { outStream, outStreamPromise } = await this.createWriteOutStream(
       path,
       mimetype,
     );
@@ -118,6 +121,17 @@ export class LocalFileService {
     const prefix = 'data:image/svg+xml;base64,';
     const base64Encoded = newAvatar.base64.slice(prefix.length);
     const buff = Buffer.from(base64Encoded, 'base64');
-    return this.saveFileDataFromBuffer(buff, AVATARS_PATH, 'image/svg+xml');
+    try {
+      return await this.saveFileDataFromBuffer(
+        buff,
+        AVATARS_PATH,
+        'image/svg+xml',
+      );
+    } catch (err) {
+      if (err instanceof Error) {
+        this.logger.error(err.message);
+      }
+      return null;
+    }
   }
 }

--- a/transcendence-app/swagger-spec/swagger-spec.yaml
+++ b/transcendence-app/swagger-spec/swagger-spec.yaml
@@ -46,6 +46,8 @@ paths:
           description: Bad Request
         "422":
           description: Unprocessable Entity
+        "503":
+          description: Service Unavailable
       tags: *a1
   /api/v1/auth/local/login:
     post:
@@ -515,6 +517,8 @@ paths:
           description: Forbidden
         "422":
           description: Unprocessable entity
+        "503":
+          description: Service Unavailable
       tags: &a5
         - chat
     get:

--- a/webapp/src/shared/generated/.openapi-generator/FILES
+++ b/webapp/src/shared/generated/.openapi-generator/FILES
@@ -1,4 +1,3 @@
-.openapi-generator-ignore
 apis/AuthApi.ts
 apis/AvatarApi.ts
 apis/ChatApi.ts

